src/dune_rules/findlib.ml

```diff
@@ -57,14 +58,22 @@ module DB = struct
       ]
   ;;
 
-  let equal t { stdlib_dir; paths; builtins; ext_lib } =
+  let equal t { stdlib_dir; paths; builtins; ext_lib; hash = _ } =
     Path.equal t.stdlib_dir stdlib_dir
     && List.equal Path.equal t.paths paths
     && Package.Name.Map.equal ~equal:Meta.Simplified.equal t.builtins builtins
     && String.equal t.ext_lib ext_lib
   ;;
 
-  let hash { stdlib_dir; paths; builtins; ext_lib } =
+  let equal a b =
+    (* Since the DB is cached per context, physical equality will
+       shortcut almost all equality tests. *)
+    phys_equal a b || equal a b
+  ;;
+
+  let hash t = t.hash
```

Comment:
Isn't caching the hash function a separate optimzation? Can you split that into its own PR please?

Reply:
In my mind it was part of the same optimization to speed up the hashtable, but sure I can split it in a separate PR.